game-core/src/main/java/games/strategy/engine/lobby/client/login/LobbyServerProperties.java

```diff
@@ -29,7 +30,7 @@
   @Nonnull private final String host;
 
   /** The port the lobby is listening on. */
-  private final int port;
+  @Nonnull private final Integer port;
 
   /** URI for the http lobby server. */
   @Nullable private final URI httpServerUri;
```

Comment:
The suggestion to favor `Optional` over a nullable type in the getter return type was motivated by the getter for this field possibly returning `null`.  I think there's a problem below on line 41:

https://github.com/triplea-game/triplea/blob/f43cfdd8d71e9e73d9e76f90c334925b8deac1f0/game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportConfiguration.java#L35-L42

Just because the `Optional<LobbyServerProperties>` is present, doesn't mean that there will be a non-`null` `httpServerUri`.  It looks like, if a `null` URI is passed to `ErrorReportClientFactory#newErrorUploader()`, it will trigger an NPE.

Maybe something like the following will work:

```diff
diff --git a/game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportConfiguration.java b/game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportConfiguration.java
index 1779b9232..b6baa593d 100644
--- a/game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportConfiguration.java
+++ b/game-core/src/main/java/games/strategy/debug/error/reporting/ErrorReportConfiguration.java
@@ -9,6 +9,7 @@ import javax.swing.JFrame;
 import org.triplea.http.client.error.report.ErrorReportClientFactory;
 
 import games.strategy.engine.lobby.client.login.LobbyPropertyFetcherConfiguration;
+import games.strategy.engine.lobby.client.login.LobbyServerProperties;
 import games.strategy.triplea.settings.ClientSetting;
 
 class ErrorReportConfiguration {
@@ -35,9 +36,7 @@ class ErrorReportConfiguration {
   private static Optional<BiConsumer<JFrame, UserErrorReport>> newFromRemoteProperties() {
     return LobbyPropertyFetcherConfiguration.lobbyServerPropertiesFetcher()
         .fetchLobbyServerProperties()
-        .map(
-            props ->
-                new ErrorReportUploadAction(
-                    ErrorReportClientFactory.newErrorUploader(props.getHttpServerUri())));
+        .flatMap(LobbyServerProperties::getHttpServerUri)
+        .map(uri -> new ErrorReportUploadAction(ErrorReportClientFactory.newErrorUploader(uri)));
   }
 }
diff --git a/game-core/src/main/java/games/strategy/engine/lobby/client/login/LobbyServerProperties.java b/game-core/src/main/java/games/strategy/engine/lobby/client/login/LobbyServerProperties.java
index a1594878d..438307d38 100644
--- a/game-core/src/main/java/games/strategy/engine/lobby/client/login/LobbyServerProperties.java
+++ b/game-core/src/main/java/games/strategy/engine/lobby/client/login/LobbyServerProperties.java
@@ -42,6 +42,10 @@ public final class LobbyServerProperties {
    */
   @Nullable private final String serverMessage;
 
+  public Optional<URI> getHttpServerUri() {
+    return Optional.ofNullable(httpServerUri);
+  }
+
   public Optional<String> getServerMessage() {
     return Optional.ofNullable(Strings.emptyToNull(serverMessage));
   }
```

Reply:
Http host URI is required now (over the course of the last few PRs); just as much as 'host' is required, NPE handling should not be necessary.

Reply:
Maybe I'm missing something obvious, but I don't see where `httpServerUri` is now required.  It's marked `@Nullable`, so the builder won't throw an exception at runtime if it's not specified.  More importantly, the factory methods `getTestOverrideProperties()` and `getLastUsedProperties()` in `LobbyServerPropertiesFetcher` never specify this URI.

For example, if I override the lobby host/port (but not the HTTP server URI), and then attempt to send an error report, I receive the following error:

![send-error-report-failure](https://user-images.githubusercontent.com/4826349/50739117-8204ab00-11a9-11e9-9314-c0de9b8c39ba.png)

That empty message is not very helpful, so we should probably send the stack trace of the exception to the log.  Regardless, stepping in with the debugger, the root cause of the failure is indeed an NPE:

```
java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:877)
	at org.triplea.http.client.FeignFactory.build(FeignFactory.java:36)
	at org.triplea.http.client.HttpClient.apply(HttpClient.java:43)
	at org.triplea.http.client.ServiceClient.apply(ServiceClient.java:23)
	at games.strategy.debug.error.reporting.ErrorReportUploadAction.sendErrorReport(ErrorReportUploadAction.java:41)
	at games.strategy.debug.error.reporting.ErrorReportUploadAction.lambda$1(ErrorReportUploadAction.java:37)
	at games.strategy.engine.framework.ui.background.BackgroundTaskRunner.lambda$0(BackgroundTaskRunner.java:51)
	at games.strategy.engine.framework.ui.background.BackgroundTaskRunner$1.doInBackground(BackgroundTaskRunner.java:129)
	at javax.swing.SwingWorker$1.call(SwingWorker.java:295)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at javax.swing.SwingWorker.run(SwingWorker.java:334)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

(To raise that NPE sooner, we probably want to mark `HttpClient#hostUri` as `@Nonnull` so the `@RequiredArgsConstructor` will throw immediately instead of deferring until the connection is attempted.)

Reply:
I wonder if i was looking at some uncommitted code, you're right @ssoloff ; some more updates are needed..